kep/108-Memory-bandwidth-and-LLC-aware-filtering-and-scoring/README.md

```diff
@@ -0,0 +1,145 @@
+# Memory bandwidth and LLC aware filtering and scoring
+
+## Table of Contents
+
+<!-- toc -->
+- [RDT aware filtering and scoring](#RDT-aware-filtering-and-scoring)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Terms](#terms)
+  - [Proposal](#proposal)
+  - [Design Details](#design-details)
+    - [Extension points](#extension-points)
+      - [Filter](#filter)
+      - [PreScore](#prescore)
+      - [Score](#score)
+  - [Known Limitations](#known-limitations)
+  - [Alternatives considered](#alternatives-considered)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Testing Plan](#testing-plan)
+  - [Implementation History](#implementation-history)
+  - [References](#references)
+<!-- /toc -->
+
+## Motivation
+
+Memory bandwidth and LLC (last level cache) are very important resources in a system. Kuberenetes scheduler hasn't taken them into consideration yet. But memory bandwidth and LLC contention may impact system performance very much. Furtherly, it causes bad SLA. Intel Resource Director Technology (RDT) is one example of the technology that can be leveraged to mitigate the impact of memory bandwidth and LLC contention. It is very helpful to ensure system performance stability and improve SLA. 
+
+## Goals
+1. Use scheduler plugin, which is the most Kubernetes native way, to implement memory bandwidth and LLC aware filtering and scoring.
+2. Leverage memory bandwidth and LLC relative metrics for filtering and scoring.
+3. Provide configurable weigths for prioritizing the metrics used in the scoring calculations.
+
+## Non-Goals
+
+## Use Cases
+
+## Terms
+LLC: last level cache.
+RDT: Intel Resource Director Technology.
+SLA: Service Level Agreement.
+MPKI: Misses Per Kilo Instructions.
+
+## Proposal
+We introduce some metrics such as memory bandwidth utilization, LLC occupation etc. At schedule stage, we can leverage these metrics to assist filting and scoring. It is hard to determine how important a resource is and how much resource is required. We did off-line analysis to different scenario and get a series of coefficient of correlation and generic usage value. The coefficient is called affinity and the generic usage value is called profile. The affinities and profiles are provided by a yaml configuration file. Detail about how the affinity and profile are used is described in the following design details section. The average value of a period of time is more meaningful than immediate value for some metrics, for instance, LLC occupancy and CPU usage. Metrics data aren't obtained from agents (such as cAdvisor) directly, instead they are stored in Prometheus first and then are retrieved on demand.
```

Comment:
> The affinities and profiles are provided by a yaml configuration file.
---
How is this yaml generated? Can users modify it or fixed and unchanging？

Reply:
It is based on off-line analysis to different usage scenario. yes, user can change it.

Reply:
> It is hard to determine how important a resource is and how much resource is required.

It's true but what if users know their workloads pretty well and already set the resource requests properly? Or, are you implying that this feature works for the workloads that don't set Pod requests explicitly? (i.e., BestEfforts pods)

Reply:
yes,  Normally memory bandwidth, LLC ... are not set in Pod requests.